### PR TITLE
onDateIsEnabled callback to determine if date is selectable

### PR DESCRIPTION
**Pull request by aforty**
I needed this for a project I'm working on. Basically I needed only past dates to be selectable and for future dates to not even be clickable. I implemented this by providing the component with a callback to check whether a specific date is enabled. If it isn't enabled then it won't be wrapped in a `<TouchableWithoutFeedback />` and a new `disabledDayText` style is applied.

This may not fit with your plans for this component but like I said, I needed this for a project and decided to implement it this way.

Also did some cleanup, for example `CalendarIOS` still had `_dayTextStyle()` and `_dayCircleStyle()` methods which were going unused. I assumed it was old code and removed it.

### CODE REVIEW

**Comment by tomaskulich** [comment]
Wow, I like this. IMO that's really nice feature.

**Comment by tomaskulich** [comment]
Also, @christopherdro, maybe you can remove the old versions of `node` from the CI process? I don't think that node v. 0.6 and 0.8 bothers anyone anymore :)

**Comment by christopherdro** [comment]
Yup I already have on master

**Comment by christopherdro** [comment]
This needs to be rebased because of the file changes as well

**Comment by tomaskulich** [comment]
@aforty Could you please re-implement the features on the current master? (I believe, this will be the least painful way of merging) Or are you currently too busy to work on this?

**Comment by evollu** [comment]
any updates on this?

**Comment by aforty** [comment]
Refactorized.

File: example/package.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
   },
   "dependencies": {
     "react": "^15.1.0",
-    "react-native": "^0.27.0-rc2",
-    "react-native-calendar": "github:christopherdro/react-native-calendar#revamp"
+    "react-native": "^0.27.0",
+    "react-native-calendar": "file:../"
```
Comment (aforty) [inline example/package.json:11]: Updated `react-native` to latest and also pulling `react-native-calendar` from parent directory, makes things easier to debug. When making changes in parent be sure to `npm un react-native-calendar && npm i` to remove previous version and to pull in fresh from parent directory.

**Comment by christopherdro** [comment]
@coderdave Thoughts?

**Comment by coderdave** [comment]
@christopherdro I like the idea of non-selectable dates. But maybe just pass in an array of dates that aren't selectable. So maybe another property called `disabledDates`? To me this would feel more natural and more in line to the current API since we already have one that is called `eventDates` that accepts an array.

**Comment by aforty** [comment]
A callback is better imho because that way you're not pre-calculating all the possible scenarios years in advance, instead the calculation is made when those months are actually rendered. This worked better in my project at the time. That said I think `eventDates` is flawed in the same way. It probably even ought to accept a promise from a callback in order to resolve async scenarios. 

Also, could you not have thought to check with other people before I refactored the code? Was the refactor necessary to consider the merits of this PR? I had to get the project setup and spent an hour+ on it. Don't accept this PR based on that but please consider other people's time in the future before asking for a refactor you don't any intention of using.

**Comment by tomaskulich** [comment]
@aforty @christopherdro @coderdave 

To avoid more wasted work, we should decide if we want to pass eventDates and disabledDates as functions (possibly async?) or as arrays.

Pros and cons for both the options:
- we should choose one consistent way how to do both eventDates and disabledDates.
- arrays (if sparse) can be implemented much faster than functions (however, this probably does not matter?)
- functions are more general (no need to limit user to some specific date range and precalculate dates for that)
- async functions seem quite useless to me: doing ~30 async calls to render just one month seems really inefficient. 
-  having array of dates, it's easy to convert this to predicate (function). It's relatively easy to do this also other way round, however you'd need to materialize the predicate on some date range (say +-1 year from the current date).
- theoretically, we can support both array and function, and make a decision based on the type of the argument. However, since the conversion is quite easy, the added value of this is questionable.

Am I missing something?

My preference:  if it does not significantly degrade the speed, change both eventDates and disabledDates to predicates (functions) accepting standard js Date object.

**Comment by christopherdro** [comment]
@tomaskulich How would you pass multiple dates as date objects without using an array?
I think the idea of using arrays has to stay but I do prefer using JS Date over strings. So that could be something we change.

@aforty I understand your frustration but please do understand that I have multiple open source projects that I have to find time to attend to, especially in times that I'm busy with actual paid work and the amount of hours you've saved by not having to write a calendar from scratch.

@coderdave @tomaskulich The next weeks are pretty intense for me so if you guys can come up with a good solution that everyone is happy with that would be awesome!!

**Comment by tomaskulich** [comment]
@christopherdro `eventDates` can be implemented the same way as @aforty did `disabledDates`: `eventDates` (probably we should rename it to something such as `eventExists` can be a predicate (function) that accepts the given date and returns true / false according to whether the event exists for the day.

**Comment by christopherdro** [comment]
How would you then specify the evenDates that should exist?

**Comment by aforty** [comment]
@christopherdro You're right, my frustration got the best of me.

Perhaps I'm thinking about this all wrong. Maybe an array of dates is actually preferable, for both `eventDates` and `inactiveDates`. The `onTouchPrev` and `onTouchNext` events can be used to further populate past/future inactive and event dates, then when those become available they will be passed down to the calendar as props. That seems like the react way of doing things.

Then perhaps both `eventDates` and `inactiveDates` could also be supplied as a callback to perform the logic at render vs ahead of time, but make this an option for the user.

So basically I'm saying: array props and callbacks are good (give the user the choice), but forget the async promises.

**Comment by tomaskulich** [comment]
When rendering a given month, the calendar will call `eventDates` (or however we rename it) 30  times (number of days in month), each time with different date as an argument. If the function returns true, it means, that the event takes place on a given day. Otherwise, there is no event for the day.

**Comment by christopherdro** [comment]
@aforty I really like that approach!

@tomaskulich What do you think about that? What we can do also after getting this merged in is accept date objects in the array but still provide backwards compatibility if a string is being used.

**Comment by aforty** [comment]
I would additionally move to make `moment` a peer dependency so that we can pass dates for the callbacks as `moment` objects. I just find `moment` to be easier to manipulate and compare than native `Date`, and that's exactly what would happen within these callbacks. 

Also perhaps we can add a `format` string which can then be used to parse incoming strings in arrays. So you could pass `eventDates={['2016-06-19']}` and `format="yyyy-MM-DD"` and have it parse correctly regardless of how you decide to pass your dates in. `eventDates` can also consist of native `Date` or `moment` objects, which would then not be subject to a `moment.parse()` using the `format` string. All this would just to increase flexibility while using the component.

**Comment by coderdave** [comment]
What does having the `eventDates` (or `inactiveDates`) to optionally take a function do for us when all you have to do is write a function that returns an array of dates?

**Comment by aforty** [comment]
@coderdave Just flexibility. Perhaps you have a very large dataset and would prefer not to calculate all the dates ahead of time when the component (I believe) only renders +/- 3 months. However I guess one could use `onTouchNext` to then calculate a new set of dates for `eventDates` and `inactiveDates`.

**Comment by coderdave** [comment]
Yes, that's what I do now is hook into onTouchNext/Previous to get my new set of dates.

**Comment by aforty** [comment]
Yeah so what I'm saying is that I'm fine leaving the callback out, but wanted to leave in as flexibility. It costs nearly nothing to leave it in as an option.

**Comment by tomaskulich** [comment]
+1 to optional callback. 

@aforty if you have some mana left, it would be nice, if you can make it such that `eventDates` works the same way.

As for the string / moment / Date, I prefer Date. It's the canonical way how to communicate date in js world. Peer dependencies are shady concept and also, you probably don't want user to install moment.js if they want to use the package without this feature.

**Comment by aforty** [comment]
@tomaskulich Fair point about peerDependencies. How about the optional date string format for passing in dates as strings?

**Comment by christopherdro** [comment]
@tomaskulich I agree that date would be best for event dates but can make things a bit more problematic when trying to change the calendars functionality to use pure dates since since we use a good amount of helper methods from moment. It is doable though of course!

The only thing I wanted to point out is that with either decision we should still provide backwards compatibility for strings as event dates and move away from them in the future.

**Comment by tomaskulich** [comment]
I see, supporting strings for backward compatibility makes sense.

**Comment by aforty** [comment]
I don't think string support should be removed at all. API's will likely provide strings, perhaps just in a different format, in which case the optional `format` prop I mentioned above could be helpful. Requiring `Date` objects adds an additional step. I think the `eventDates` and `inactiveDates` should accept either a callback method (which get passed a native `Date`) or an array of `String`, `Date` or `moment` objects. If an array element is a string it is parsed using `moment(dateString, format)` where `format` can be passed in as a prop but which defaults to YYYY-MM-DD.

**Comment by christopherdro** [comment]
@aforty Any chance you still feel like adjusting to the latest comments so we can merge it in?

**Comment by aforty** [comment]
I can still get to this at some point, I've just been really busy lately.

**Comment by davorbadrov** [comment]
Any updates on this? Is it going to be accepted?

**Comment by pbassut** [comment]
There's some conflicts on this. 
Will this get merged anytime soon?

**Comment by aforty** [comment]
Wow, almost a year since I said I'd try to get to this. Not sure if it's going to happen anymore, I haven't used this component since then and I'm no longer involved in the project that I used it on initially. So I'm pretty far removed from this code base and I'm not sure if I'd be the best person to try and refactor this.

Closing.

**Comment by batistafrancis** [comment]
Is this was considered?
